Design/Instrument-2.0/ScanningInstruments_Design.md

```diff
@@ -0,0 +1,170 @@
+
+```cpp
+for (size_t i = 0; i < numberOfTubes; ++i) {
+  for (size_t j = 0; j < m_numberOfPixelsPerTube; ++j) {
+    int *data_p = &data(static_cast<int>(i), static_cast<int>(j), 0);
+    m_localWorkspace->setHistogram(spec, m_localWorkspace->binEdges(0),
+        Counts(data_p, data_p + m_numberOfChannels));
+    spec++;
+  }
+}
+```
+
+This will have performance impacts for the loader, but can also leave the workspace in an inconsistent state. For the scanning instrument design the workspace should be initialised all at once, via a `DetectorInfo` factory, which will ensure the loader creates a workspace in a consistent state.
+
+### Design
+
+#### `DetectorInfo`
+
+The core of the changes to support scanning instruments will be in the [`DetectorInfo`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/DetectorInfo.h) class. There are two choices for implementing the required changes in `DetectorInfo`. Currently the methods `isMonitor`, `isMasked`, `l2`, `twoTheta`, `signedTwoTheta`, `position`, `rotation`, `setPosition` and `setRotationEach` are accessed by a detector index. This could be kept and the detector index would be associated with a detector ID and a time index. Alternatively the methods can be accessed by supplying the detector index, which has a one-to-one correspondence with a detector ID, and a time index. There would be a time index corresponding to each position in the step scan.
+
+The current plan is to access to add methods that take a detector index and a time index, in addition to the methods that take just a detector index. Accessing a scanning instrument with just a detector index would throw. This will save memory on detector metadata which would not change for different time indices, such as the monitor flag. See the end of this document for the changes to the signature of some of the public `DetectorInfo` methods.
+
+The `timeIndex` argument is optional, so algorithms that do not care about time indexing will simply return the first/only time index. `DetectorInfo` should also contain a vector of start and end times for each time index. This will mean any information stored as a time series sample log can also be obtained for the correct step, for example the monitor counts for normalisation.
+
+`DetectorInfo` will ultimately store the information for positions and rotation, so will need to do this in a way that can be accessed by both the detector index and time index. Masking would be done at the spectrum level, if it is required for a detector for some time steps, but not all. `DetectorInfo` will also hold the start and end times for each step in the scan. The objects it will need to hold for this would be:
+* `std::vector<std::vector<Kernel::V3D>> positions`
+* `std::vector<std::vector<Kernel::V3D>> rotations`
+* `std::vector<std::vector<std:pair<Kernel::DateAndTime, Kernel::DateAndTime>>> timeRanges`
```

Comment:
I was thinking of doing this a bit like this, but end the end we will just have to see what works best, maybe leave this more open in the design document?
https://github.com/DMSC-Instrument-Data/instrument-prototype/pull/38#issuecomment-263863475